nixos/paperless: refactor to use systemd LoadCredential, Switch to systemd.tmpfiles.settings

Ramblurr opened the pull request:
## Description of changes

This PR contains two commits

1. > nixos/paperless: refactor to use systemd LoadCredential
    
    This replaces the paperless-copy-password service with the use of
    systemd's LoadCredential mechanism.
    
    It is not a breaking change since it is gated behind `cfg.passwordFile`.

3. > nixos/paperless: Add opt `services.paperless.createDirs`
    
    This option resolves #301746 by allowing the admin to bypass the
    creation of the paperless directories by systemd-tmpfiles.
    
    This is necessary when, for example, those directories lie inside an NFS
    mount that the root user does not have rw access to.

    It is also not a breaking change because the default behavior is the old behavior.


Both changes are covered in the test suite.

@Flakebi
@leona-ya
@erikarvstedt

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Ramblurr committed:
nixos/paperless: refactor to use systemd LoadCredential

This replaces the paperless-copy-password service with the use of
systemd's LoadCredential mechanism.

It is not a breaking change since it is gated behind `cfg.passwordFile`.

ambroisie commented on nixos/tests/paperless.nix:
```diff
@@ -85,5 +91,13 @@ import ./make-test-python.nix ({ lib, ... }: {
+    postgres.wait_for_shutdown()
+
+    with subtest("When directory creation is disabled, the service should fail to start"):
+        noCreateDirs.sleep(10)
```
Writing your own `wait_for_failed_unit` (based on `wait_for_unit`) should work well enough?

Ramblurr replied:
Should I add it to the shared wait_for helpers?

ambroisie replied:
I'll defer to owners of the test runner for this :shrug:.

Ramblurr replied:
Added.

SuperSandro2000 commented on nixos/modules/services/misc/paperless.nix:
```diff
@@ -134,6 +134,17 @@ in
       description = lib.mdDoc "Whether all users can write to the consumption dir.";
     };
 
+    createDirs = mkOption {
```
I would recommend instead to use mkForce for the specific paths in tmpfilesd as this option is very specific to your exact setup and not generally useful IMO.

erikarvstedt replied:
I guess what @SuperSandro2000 means is to convert `tmpfiles.rules` to `tmpfiles.settings`, like so:

```nix
systemd.tmpfiles.settings."10-paperless" = let
  defaultRule = {
    inherit (cfg) user;
    inherit (config.users.users.${cfg.user}) group;
  };
in {
  "${cfg.dataDir}".d = defaultRule;
  "${cfg.mediaDir}".d = defaultRule;
  "${cfg.consumptionDir}".d = if cfg.consumptionDirIsPublic then { mode = "777"; } else defaultRule;
};
```

Then you can use the module system's builtin features to disable dir creation:
```nix
systemd.tmpfiles.settings."10-paperless" = mkForce {};
```

Yes, this approach is better than an extra option!

Ramblurr replied:
great, thanks for the feedback. I'll refactor the PR

Ramblurr replied:
Refactored, thanks for the feedback.

erikarvstedt reviewed (changes requested):
Test `When directory creation is disabled, the service should fail to start` can be removed.
This property trivially holds true for all services that use dirs created via `systemd.tmpfiles` and doesn't need to be tested.

Ramblurr committed:
nixos/paperless: Switch to `systemd.tmpfiles.settings`

This option resolves #301746 by allowing the admin to bypass the
creation of the paperless directories by systemd-tmpfiles.

This is necessary when, for example, those directories lie inside an NFS
mount that the root user does not have rw access to.

Fixes #301746

Ramblurr commented:
> Test `When directory creation is disabled, the service should fail to start` can be removed. This property trivially holds true for all services that use dirs created via `systemd.tmpfiles` and doesn't need to be tested.

Test removed!

SuperSandro2000 commented on nixos/modules/services/misc/paperless.nix:
```diff
@@ -220,15 +220,16 @@ in
-    ];
+    systemd.tmpfiles.settings."10-paperless" = let
+      defaultRule = {
+        inherit (cfg) user;
```
```suggestion
```
seems to be unused

erikarvstedt commented:
Can you rebase this to master to fix `nixosTests.paperless`?
Currently, dependency `python3.11-moto-5.0.3` fails to build.

Ramblurr commented:
Done.

erikarvstedt commented:
The `moto` build failure only appeared when building `nixosTests.paperless` with https://github.com/NixOS/nix/pull/6530. It was unrelated to the rebase.

erikarvstedt reviewed (approved):
Nice work!

- [x] `nixosTests.paperless` succeeds on `x86_64-linux`

leona-ya reviewed (approved):
This is nice, thank you!
